src/hardware/mixer.cpp

```diff
@@ -120,7 +118,17 @@ struct mixer_t {
 
 static struct mixer_t mixer = {};
 
-uint8_t MixTemp[MIXER_BUFSIZE] = {};
+alignas(sizeof(void *)) uint8_t MixTemp[MIXER_BUFSIZE] = {};
```

Comment:
Nice! This will surely help the compiler SIMDify more of the the simple loops that access this buffer.

In the PC Speaker, we do: `int16_t *buffer = reinterpret_cast<int16_t *>(MixTemp);`

Which is a bit dangerous because it assumes/relies on the original buffer being at-least aligned to 2-byte boundaries, so at a minimum we should be `alignas(int16_t)`'ing it, which your addition takes care of going up to the pointer size.

From what I can see, we should align buffer to the largest type we plan to write into it (if we're still going the reinterpret_cast-route).  I think that's 32bit (either float or int) -- Maybe we could use `alignas(float)`?

Reply:
@kklobe , thinking about the other big performance-intensive memory blocks into which we write varying types, would a similar approach be useful?

If we're thinking that `alignas(sizeof(void *))` could be useful in lots of places, maybe we could put that boiler-plate in `mem.h` or `types.h`, and then name it something like `align_buffer` or `align_mem`?

Reply:
@kcgen this is an interesting efficiency concept - what is the "best" alignment, or at least not the worst (i.e. single-byte). Without going off into the weeds, I suppose if we wanted to do the best we can for `MixTemp` and other larger buffers, we should go for system page alignment?

```
#ifndef PAGESIZE
#define PAGESIZE 4096
#endif

alignas(PAGESIZE) uint8_t MixTemp[MIXER_BUFSIZE] = {};
```

Reply:
update: although it's looking like the compiler is doing it without the alignas, even.

Reply:
Nope, it seems to be random with no `alignas` with frequent byte-alignments, carry on with the `PAGESIZE` suggestion!

Reply:
> this is an interesting efficiency concept - what is the "best" alignment, or at least not the worst (i.e. single-byte). Without going off into the weeds

Yeah, I think that's where we probably should just leave the compiler to decide. We should provide a minimum and the compiler can ratchet up if it things there's benefits.

With GCC, you can control this at the compiler-phase without having to riddle the code with align sizes. Stuff like: `-falign-functions=32 -falign-jumps=32:8:8 -falign-loops=32:25:8 ...etc ` 

And the alignments details can get extremely processor-specific:

---

-falign-functions
-falign-functions=n
-falign-functions=n:m
-falign-functions=n:m:n2
-falign-functions=n:m:n2:m2

Align the start of functions to the next power-of-two greater than or equal to n, skipping up to m-1 bytes. This ensures that at least the first m bytes of the function can be fetched by the CPU without crossing an n-byte alignment boundary.

If m is not specified, it defaults to n.

  Examples: -falign-functions=32 aligns functions to the next 32-byte boundary, -falign-functions=24 aligns to the next 32-byte boundary only if this can be done by skipping 23 bytes or less, -falign-functions=32:7 aligns to the next 32-byte boundary only if this can be done by skipping 6 bytes or less.

The second pair of n2:m2 values allows you to specify a secondary alignment: -falign-functions=64:7:32:3 aligns to the next 64-byte boundary if this can be done by skipping 6 bytes or less, otherwise aligns to the next 32-byte boundary if this can be done by skipping 2 bytes or less. If m2 is not specified, it defaults to n2. 

---

Fortunately when building against a base architecture `-march=` type, the compiler is populated with these code/size/alignment modules, and will choose the best ones (or good defaults).

With Gentoo, I remember playing with these values and it was surprising how quickly I could destroy performance setting everything to `=32` :laughing: .  I soon learned to never touch them again and leave them defaults!

In this case, we're specifying an 8-bit array but actually abusing it by holding up-to-32bit values, so we definitely need 32-bit alignment.  We need that alignment no matter what to avoid undefined behavior (let alone worrying about optimizations).

If we can get measurable speedups by going beyond that, especially for the major hot-path arrays, then sure, we can let the data justify the need for extra alignment.

Maybe there's an easy way we can test it with unit tests?
 - Allocate a handful of side-by-size arrays, some with default (unprescribed) alignment, and then ratchet up and see if some micro-benchmark changes?

Reply:
> In this case, we're specifying an 8-bit array but actually abusing it by holding up-to-32bit values, so we definitely need 32-bit alignment. We need that alignment no matter what to avoid undefined behavior (let alone worrying about optimizations).

Um, I'm with @kcgen and I don't think we should screw around with this any stuff at such low level, except for aligning byte-buffers to the native width, and that's about it.

Reply:
I'd be wary about micro benchmarks in general: yes, you can achieve maybe 10 to 100 fold performance increase in a few hotspots, but unless they also take up significant time of the total run time, it's a bit of a wasted effort (in the sense that the total CPU consumption can remain unchanged). Plus some optimisations might look excellent on paper in isolation, but might as well introduce problems elsewhere (e.g. ruin cache performance in other parts of the app).

Personally, I'd focus on other improvements than performance optimisations (unless they're a big problem); I think that would be energy and time better spent, and we don't have unlimited amounts of that, unfortunately 😁 

Of course, I'm not trying to prevent anyone from optimising the performance if they're super into this, just some general remarks.

Reply:
On the related tangents front: @kklobe 's previously profiled the memory accessors (functions in _mem* headers) soak up a significant portion of the total runtime; so there are surely gains to be made. He'll find them 😎

Reply:
> In this case, we're specifying an 8-bit array but actually abusing it by holding up-to-32bit values, so we definitely need 32-bit alignment. We need that alignment no matter what to avoid undefined behavior (let alone worrying about optimizations).

Will those compiler options align byte arrays? I wasn't sure from your description, but even if so, rather than relying on compiler-specific switches for alignments, the standard keyword seems helpful. And, if we have to specify a value anyways, system pagesize just hits all the marks. Unlikely to move the needle at all from the raw performance benchmark perspective, of course!

Reply:
> Personally, I'd focus on other improvements than performance optimisations (unless they're a big problem); I think that would be energy and time better spent, and we don't have unlimited amounts of that, unfortunately 😁

Agree. I generally make these easy cleanups when I come across them while doing something else; in this case, spelunking around in the mixer to see what was happening with the AdLib Gold issue in Dune.

Staging has come as far as it has not only due to great new features but also the occasional weed-pulling here and there to get away from "death by a thousand paper cuts", and a good number of our users are running older setups. Sometimes a little janitorial work is a nice break from the harder stuff.

Thanks for the feedback, and thanks especially for all the audio work (I'm really enjoying the SB improvements)!

Reply:
Sure, cleaning up stuff always helps, no doubt about that. I guess I'm just coming from the perspective that every single DOS game runs under ~5% CPU on my 7 year old machine, so my focus is on stability and features, that's all.

Glad you're enjoying the audio improvements, and hints about the mysterious Dune issues are always welcome if you spotted something fishy 😅

Reply:
>  my focus is on stability and features, that's all.
> 
Please, keep doing that!

Reply:
> Will those compiler options align byte arrays? I wasn't sure from your description,

After testing them, atleast on x86_64, I can only get alignment at the type-size level. 

``` c++
#include <cstdio>
#include <cstdint>

int main(void)
{
  const uint8_t a[1024] = {};
  const int16_t b[3] = {};
  const int64_t d[3] = {};
  const uint32_t c = {};
  const char e[] = "testing";

  printf("uint8_t a[1024]: %p, size %zu, alignment %zu\n", (void*)&a, sizeof(a), alignof(a));
  printf("int16_t b[3]   : %p, size %zu, alignment %zu\n", (void*)&b, sizeof(b), alignof(b));
  printf("int64_t d[3]   : %p, size %zu, alignment %zu\n", (void*)&d, sizeof(d), alignof(d));
  printf("uint32_t c     : %p, size %zu, alignment %zu\n", (void*)&c, sizeof(c), alignof(c));
  printf("char e[]       : %p, size %zu, alignment %zu\n", (void*)&e, sizeof(e), alignof(e));

  return 0;
}
```

For example:

```text
[10:59:21@server alignment] g++ -mtune=native -O3 -falign-functions=32 -falign-jumps=32 -falign-loops=32 -falign-labels=32 align_test.cpp 
[10:59:26@server alignment] ./a.out 
uint8_t a[1024]: 0x7ffc5b93a3c0, size 1024, alignment 1
int16_t b[3]   : 0x7ffc5b93a39a, size 6, alignment 2
int64_t d[3]   : 0x7ffc5b93a3a0, size 24, alignment 8
uint32_t c     : 0x7ffc5b93a394, size 4, alignment 4
char e[]       : 0x7ffc5b93a3b8, size 8, alignment 1
```

> but even if so, rather than relying on compiler-specific switches for alignments

Definitely agree; the code should be correct on it's own merit.

I think all those alignment-tweaks involve cost-models and optimizations.

Here's another couple:

```
align-loop-iterations

    A loop expected to iterate at least the selected number of iterations is aligned.

vect-max-peeling-for-alignment

    The maximum number of loop peels to enhance access alignment for vectorizer. Value -1 means no limit.

```

It looks like the compiler (at all optimization levels) keeps the scalars and arrays aligned at least to their type-size.